File: 2015/go/the-ideal-stocking-stuffer/main.go
package main

import (
	"crypto/md5"
	"fmt"
	"math"
)

const Input = "bgvyzdsv"

func part1(input string) int {
	for i := 1; i < math.MaxInt; i++ {
		hash := md5.Sum([]byte(fmt.Sprintf("%s%d", input, i)))
		if hash[0] == 0 && hash[1] == 0 && hash[2]&0xf0 == 0 {
			return i
		}
	}
	return 0
}

func part2(input string) int {
	for i := 1; i < math.MaxInt; i++ {
		hash := md5.Sum([]byte(fmt.Sprintf("%s%d", input, i)))
		if hash[0] == 0 && hash[1] == 0 && hash[2] == 0 {
			return i
		}
	}
	return 0
}

func main() {
	println("Part 1:", part1(Input))
	println("Part 2:", part2(Input))
}

File: 2015/js/the-ideal-stocking-stuffer/src/main.js
const { createHash } = await import('node:crypto');

const INPUT = 'bgvyzdsv';

function findPrefix(input, prefix) {
    for (let i = 1; i < Number.MAX_SAFE_INTEGER; i++) {
        const digest = createHash('md5').update(`${input}${i}`).digest('hex');
        if (digest.startsWith(prefix)) {
            return i;
        }
    }
    return undefined;
}

function part1() {
    return findPrefix(INPUT, '00000')
}

function part2() {
    return findPrefix(INPUT, '000000')
}


export { part1, part2 };